File: tools/inventory.helper.js
exports.cookInventoryData = function (rawInventoryData) {
  // cache variable.
  const productNoIdxMap = {};
  let count = 0;
  const result = [];
  // step1.
  const tableHeadInfos = rawInventoryData[0].map(ele => ele.trim());
  const productNoIdx = tableHeadInfos.indexOf("商品");
  const remainingQuantityIdx = tableHeadInfos.indexOf("库存数量");
  const productNameIdx = tableHeadInfos.indexOf("品名");
  const storeNameIdx = tableHeadInfos.indexOf("店仓");
  // step2.
  rawInventoryData.forEach((element, idx) => {
    const productNo = element[productNoIdx];
    const remainingQuantity = element[remainingQuantityIdx];
    const productName = element[productNameIdx];
    // const storeName = element[storeNameIdx];
    const storeName = "可用库存"

    const mapIdx = productNoIdxMap[productNo]
    if (idx !== 0) {
      if (mapIdx !== 0 && !mapIdx) {
        productNoIdxMap[productNo] = count;
        count++;
        result.push({
          storeName, productNo, remainingQuantity, productName,
        })
      } else {
        const {storeName, productNo, remainingQuantity, productName} = result[mapIdx];
        result[mapIdx] = {
          storeName, 
          productNo, 
          remainingQuantity: result[mapIdx].remainingQuantity + remainingQuantity,
          productName,
        }
      }
    }
  });

  const productNoMap  = {}
  result.forEach((ele)=>{
    productNoMap[ele.productNo] = {
      ...ele,
    }
  })

  return productNoMap
}

File: inventory-assistant-go-service/tools/inventory.helper.go
package tools

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/thoas/go-funk"
	"github.com/xuri/excelize/v2"
)

type CookedInventoryData struct {
	StoreName         string `json:"storeName"`
	ProductNo         string `json:"productNo"`
	RemainingQuantity int    `json:"remainingQuantity"`
	ProductName       string `json:"productName"`
}

func CookInventory() {
	// read data from spreadsheet
	rawDataArr, _ := readInventorySpreadSheet()
	headRows := funk.Map((*rawDataArr)[0], func(ele string) string {
		return strings.Trim(ele, " ")
	})
	productNoIdx := funk.IndexOf(headRows, "商品")
	remainingQuantityIdx := funk.IndexOf(headRows, "可用量")
	productNameIdx := funk.IndexOf(headRows, "品名")

	// define variable
	// var result = make(map[int][]CookedInventoryData)
	// var productNoIdxMap = map[int]int{
	// 	0: -1,
	// }
	// var count = 0

	for idx, element := range *rawDataArr {
		if idx == 0 {
			continue
		}
		productNo := element[productNoIdx]
		remainingQuantity, _ := strconv.Atoi(element[remainingQuantityIdx])
		productName := element[productNameIdx]
		storeName := "可用库存"

		var cookedInventoryData = CookedInventoryData{
			ProductNo:         productNo,
			RemainingQuantity: remainingQuantity,
			StoreName:         storeName,
			ProductName:       productName,
		}
		fmt.Println("", cookedInventoryData)

	}
}
func readInventorySpreadSheet() (*[][]string, error) {
	wd, _ := os.Getwd()
	f, err := excelize.OpenFile(filepath.Join(wd, "./assets/inventory.xlsx"))

	if err == nil {
		rows, _ := f.GetRows(f.GetSheetName(0))
		return &rows, nil
	} else {
		return nil, err
	}
}
